Wrappers/Python/cil/optimisation/algorithms/PD3O.py

```diff
@@ -0,0 +1,145 @@
+        :math:`f` should be differentiable with a Lipschitz continuous gradient and :math:`A` is a bounded linear operator. 
+    
+        Parameters
+        ----------
+        f : Function
+            A smooth function with Lipschitz continuous gradient.
+        g : Function
+            A convex function with a computationally computable proximal.
+        h : Function
+            A composite convex function.
+        operator: Operator
+            Bounded linear operator
+        delta: Float, optional, default is  `1./(gamma*operator.norm()**2)`
+            The dual step-size 
+        gamma: Float, optional, default is `2.0/f.L`
+            The primal step size 
+        initial : DataContainer, optional default is a container of zeros, in the domain of the operator 
+            Initial point for the ProxSkip algorithm.             
+
+
+        Reference
+        ---------
+        Yan, M. A New Primal–Dual Algorithm for Minimizing the Sum of Three Functions with a Linear Operator. J Sci Comput 76, 1698–1717 (2018). https://doi.org/10.1007/s10915-018-0680-3
+     """    
+
+
+    def __init__(self, f, g, h, operator, delta=None, gamma=None, initial=None, **kwargs):
+
+        super(PD3O, self).__init__(**kwargs)
+
+              
+        self.set_up(f=f, g=g, h=h,  operator=operator, delta=delta, gamma=gamma, initial=initial, **kwargs)
+ 
+                  
+    def set_up(self, f, g, h, operator, delta=None, gamma=None, initial=None,**kwargs):
+        
+        logging.info("{} setting up".format(self.__class__.__name__, ))
+        
+        self.f = f # smooth function
+        if isinstance(self.f, ZeroFunction):
+            warnings.warn(" If self.f is the ZeroFunction, then PD3O = PDHG. Please use PDHG instead. Otherwise, select a relatively small parameter gamma ", UserWarning)
+            if gamma is None:
+                gamma = 1.0/operator.norm()                      
+        
+        self.g = g # proximable
+        self.h = h # composite
+        self.operator = operator
+        
+        if gamma is None:
+            gamma = 0.99*2.0/self.f.L
+        
+        if delta is None :
+            delta = 0.99/(gamma*self.operator.norm()**2)
+        
+        self.gamma = gamma
+        self.delta = delta  
+
+        if initial is None:
+            self.x = self.operator.domain_geometry().allocate(0)
+        else:
+            self.x = initial.copy()
+
+        self.x_bar = self.x.copy()    
+        self.x_old = self.operator.domain_geometry().allocate(0)
+        
+        self.s_old = self.operator.range_geometry().allocate(0)
+        self.s = self.operator.range_geometry().allocate(0)
+                
+        self.grad_f = self.operator.domain_geometry().allocate(0)        
+  
+        self.configured = True
+        logging.info("{} configured".format(self.__class__.__name__, ))
+        
+
+    def update(self):
+        r""" Performs a single iteration of the PD3O algorithm        

```

Comment:
By reordering the operations we can reduce the memory. If the proximal conjugate step happened at the end of update (and the necessary operation in `set_up` to enable it) Then `x_bar` and `x_old` won't overlap and can share the same memory buffer.

Reply:
Done